authentication/authentication.go

```diff
@@ -0,0 +1,145 @@
+package authentication
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+//nolint:gochecknoglobals
+// authenticatorFactories map is used for the providers' self-registration.
+var authenticatorFactories map[string]AuthenticatorFactory
+
+//nolint:gochecknoglobals
+// authenticatorsMtx is used to protect the authenticatorFactories.
+var authenticatorsMtx sync.RWMutex
+
+//nolint:gochecknoinits
+func init() {
+	authenticatorFactories = make(map[string]AuthenticatorFactory)
+}
+
+// onboardNewProvider is used by the providers to register themselves.
+func onboardNewProvider(providerType string, factory AuthenticatorFactory) {
+	authenticatorsMtx.Lock()
+	defer authenticatorsMtx.Unlock()
+
+	authenticatorFactories[providerType] = factory
+}
+
+// AuthenticatorFactory represents each authenticator factory, a function to create a new corresponding
+// authentication provider.
+type AuthenticatorFactory func(config map[string]interface{}, tenant string, registrationRetryCount *prometheus.CounterVec, logger log.Logger) (Provider, error)
```

Comment:
Nit: since this is a function that returns new `Providers`, maybe we can instead call this a `ProviderFactory`? It should be pretty clear what this provider is for since it's part of the `authentication` package. I think this naming would be more consistent and would reduce the cognitive overload of the difference between an `authenticator` and a `provider`, when there really is none.
```suggestion
type ProviderFactory func(config map[string]interface{}, tenant string, registrationRetryCount *prometheus.CounterVec, logger log.Logger) (Provider, error)
```

Reply:
Authenticator and Provider are used interchangeably, basically they are the same. As you mentioned, since it is under the authentication package, it would be more clear to use provider explicitly. I will update that.